prog/rand.go

```diff
@@ -754,3 +761,76 @@ func (a *CsumType) generate(r *randGen, s *state) (arg Arg, calls []*Call) {
 	// Filled at runtime by executor.
 	return MakeConstArg(a, 0), nil
 }
+
+// Finds a compatible resource with the type `t` and the calls that initialize that resource.
+func resourceCentric(t *ResourceType, s *state, r *randGen) (resource *ResultArg, calls []*Call) {
+	var p *Prog
+	for idx := range r.Perm(len(s.corpus)) {
+		p = s.corpus[idx].Clone()
+		resources := getCompatibleResources(p, t.TypeName, r)
+		if len(resources) > 0 {
+			resource = resources[r.Intn(len(resources))]
+			break
+		}
+	}
+
+	// No compatible resource was found.
+	if resource == nil {
+		return nil, nil
+	}
+
+	// Remove unrelated calls from the program.
+	relatedResources := map[*ResultArg]bool{resource: true}
+	for idx := len(p.Calls) - 1; idx >= 0; idx-- {
+		includeCall := false
+		var newResources []*ResultArg
+		ForeachArg(p.Calls[idx], func(arg Arg, _ *ArgCtx) {
+			if a, ok := arg.(*ResultArg); ok {
+				if !relatedResources[a] && len(a.uses) > 0 {
```

Comment:
I am not sure `len(a.uses) > 0` part is correct.
`len(a.uses) > 0` means that this resource is used in a subsequent call. But since we scan backwards from the last call, this is not important already.
What's important is if this resource is a use a resource created by some previous call. And I think for this we need to check `a.Res != nil`.

Reply:
It's strange no test failed...
If I am reading this correctly, it should lead to malformed programs that contain resources that are not created by any call in the program...

Reply:
I think that implementation didn't collect any related resource and these were replaced with DefaultArg. I changed the implementation to add in that map only the references to other resources. PTAL @dvyukov